tck/src/main/java/org/eclipse/microprofile/lra/tck/service/spi/LRARecoveryService.java

```diff
@@ -0,0 +1,83 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.eclipse.microprofile.lra.tck.service.spi;
+
+import java.net.URI;
+import java.util.logging.Logger;
+
+/**
+ *     The LRA model guarantees eventual consistency but does not say when
+ *     participants will be brought into a consistent state. When a request
+ *     to close/cancel an LRA is received, the implementation is allowed to
+ *     delay sending out the complete/compensate callbacks (the specification
+ *     only requires that an implementation must eventually send them).
+ *
+ *     This SPI provides a mechanism for implementations to indicate when
+ *     it knows that callbacks have been sent and responses received or when
+ *     callbacks have been issued and all participants have reached an end state.
+ *
+ */
+public interface LRARecoveryService {
+
+    /**
+     * Wait for the delivery of all participant callbacks (Complete or Compensate).
+     * When this method returns the caller can be certain that the callbacks were
+     * sent and responses received (including error responses).
+     *
+     * @param lraId the LRA context
+     * @throws LRACallbackException the implementation was unable to determine whether
+     * or not the callbacks were received by all participants
+     */
+    void waitForCallbacks(URI lraId) throws LRACallbackException;
+
+    /**
+     * Wait for all participants to reach an end state and for all
+     * {@link org.eclipse.microprofile.lra.annotation.AfterLRA} notifications to be delivered.
+     * 
+     * The default implementation iterates {@link LRARecoveryService#waitForEndPhaseReplay(URI)} until
+     * all participants reach a final state.
+     *
+     * @param lraId the LRA context
+     * @throws LRACallbackException the implementation was unable to determine whether
+     * or not all participants have reached an end state.
+     */
+    default void waitForRecovery(URI lraId) throws LRACallbackException {
+        Logger log = Logger.getLogger(LRARecoveryService.class.getName());
+        int counter = 0;
+
+        do {
+            log.info("Recovery attempt #" + ++counter);
+        } while (!waitForEndPhaseReplay(lraId));
+        log.info("LRA " + lraId + "has finished the recovery");
+    }
+
+    /**
+     * Wait for one replay of the end phase of the LRA (the calls to Status, Complete, Compensate, and Forget
+     * methods of all Compensating/Completing participants and the possible repeated calls to AfterLRA methods for
```

Comment:
The text is vague with respect to the listener callbacks - does the method perform repeated listener notifications in the same pass or does the caller need to reissue the `waitForEndPhaseReplay` repeatedly until no exception is thrown or when true is returned?